File: tests/contract/go-fixtures/rest.go
// Package main: REST route fixture capture.
//
// Each supported REST route is exercised in-process via httptest.NewRequest +
// httptest.NewRecorder so the RemoteAddr can be controlled. This is what lets
// the harness capture both the loopback auth-bypass path (success cases) and
// the non-loopback unauthenticated rejection path (401 cases) from the same
// fully-wired server, without depending on a real LAN interface.
//
// Every captured response is written to golden/rest/<name>.json as:
//
//	{
//	  "method": "GET",
//	  "path":   "/health",
//	  "status": 200,
//	  "contentType": "application/json",
//	  "body": "<redacted raw body string>"
//	}
//
// The body is the raw response body (redacted). JSON bodies are kept as raw
// text so the future Rust differential runner can choose semantic (parsed)
// comparison for object/array responses and exact byte comparison for
// contractually-significant text (error messages, markdown exports).
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
)

// restFixture is the on-disk shape of a single REST golden fixture.
type restFixture struct {
	Method      string `json:"method"`
	Path        string `json:"path"`
	Status      int    `json:"status"`
	ContentType string `json:"contentType"`
	Body        string `json:"body"`
}

// restCase describes one REST request to capture.
type restCase struct {
	// name is the golden filename (without extension).
	name string
	// method is the HTTP method.
	method string
	// path is the request path (may include query string).
	path string
	// body is the request body, or "" for no body.
	body string
	// loopback controls the request's RemoteAddr. true => "127.0.0.1:1234"
	// (auth bypass, the host-browser path); false => "10.0.0.7:1234" (a
	// remote LAN device, subject to device-credential auth).
	loopback bool
	// authHeader is an optional Authorization header value. When set on a
	// non-loopback request, the auth middleware validates it. Use a fake
	// "Bearer x:y" to exercise the 401 rejection path.
	authHeader string
	// origin is an optional Origin header. When set on a mutating loopback
	// request, the CSRF Origin check is exercised (cross-origin => 403).
	origin string
}

// captureREST runs every REST case against the harness server and writes
// golden/rest/<name>.json for each. It returns the first error encountered.
func captureREST(h *harness, goldenDir string) error {
	outDir := filepath.Join(goldenDir, "rest")
	if err := os.MkdirAll(outDir, 0o755); err != nil {
		return fmt.Errorf("mkdir %s: %w", outDir, err)
	}

	// Seed a workspace so the workspace-scoped routes have a real target. The
	// workspace ID is deterministic (hash of the absolute path), so it is
	// stable across runs and safe to reference in fixture paths.
	wsID := seedWorkspace(h)

	// Register a fake agent so /api/agents returns a populated list and
	// /api/sessions create has a target (it will still fail to launch the
	// agent binary, which is itself a contract-relevant 400).
	seedAgent(h)

	cases := buildRESTCases(wsID)
	for _, c := range cases {
		fix, err := runRESTCase(h, c)
		if err != nil {
			return fmt.Errorf("rest case %s: %w", c.name, err)
		}
		// Marshal then redact the FULL fixture text so the envelope's request
		// path (which may carry the workspace ID) is scrubbed alongside the
		// body. This keeps golden files byte-stable across runs.
		data, err := json.MarshalIndent(fix, "", "  ")
		if err != nil {
			return fmt.Errorf("marshal %s: %w", c.name, err)
		}
		redacted := h.redactor.String(string(data))
		path := filepath.Join(outDir, c.name+".json")
		if err := os.WriteFile(path, []byte(redacted+"\n"), 0o644); err != nil {
			return fmt.Errorf("write %s: %w", c.name, err)
		}
	}
	return nil
}

// buildRESTCases enumerates every supported REST route with at least a success
// case and a relevant failure case. wsID is the seeded workspace ID used in
// workspace-scoped paths.
func buildRESTCases(wsID string) []restCase {
	var cases []restCase

	// --- Health (no auth) ---
	cases = append(cases,
		restCase{name: "health_ok", method: http.MethodGet, path: "/health", loopback: true},
	)

	// --- Pairing (no auth, rate-limited) ---
	cases = append(cases,
		restCase{name: "pair_initiate_ok", method: http.MethodPost, path: "/api/pair/initiate", body: `{"host":"localhost","port":7337}`, loopback: true},
		restCase{name: "pair_initiate_bad_body", method: http.MethodPost, path: "/api/pair/initiate", body: `{not json`, loopback: true},
		restCase{name: "pair_verify_passcode_bad_body", method: http.MethodPost, path: "/api/pair/verify-passcode", body: `{not json`, loopback: true},
		restCase{name: "pair_verify_passcode_wrong", method: http.MethodPost, path: "/api/pair/verify-passcode", body: `{"passcode":"wrong wrong wrong wrong","deviceName":"dev"}`, loopback: true},
		restCase{name: "pair_verify_token_wrong", method: http.MethodPost, path: "/api/pair/verify-token", body: `{"token":"deadbeef","deviceName":"dev"}`, loopback: true},
	)

	// --- Devices (auth) ---
	cases = append(cases,
		restCase{name: "devices_list_ok", method: http.MethodGet, path: "/api/devices", loopback: true},
		restCase{name: "devices_list_unauth", method: http.MethodGet, path: "/api/devices", loopback: false},
		restCase{name: "devices_revoke_not_found", method: http.MethodDelete, path: "/api/devices/nonexistent", loopback: true},
		restCase{name: "devices_cancel_revocation_bad_body", method: http.MethodPost, path: "/api/devices/cancel-revocation", body: `{not json`, loopback: true},
	)

	// --- Pending actions ---
	cases = append(cases,
		restCase{name: "pending_actions_list_ok", method: http.MethodGet, path: "/api/pending-actions", loopback: true},
		restCase{name: "pending_actions_list_unauth", method: http.MethodGet, path: "/api/pending-actions", loopback: false},
		restCase{name: "workspaces_cancel_registration_bad_body", method: http.MethodPost, path: "/api/workspaces/cancel-registration", body: `{not json`, loopback: true},
	)

	// --- Workspaces (auth) ---
	cases = append(cases,
		restCase{name: "workspaces_list_ok", method: http.MethodGet, path: "/api/workspaces", loopback: true},
		restCase{name: "workspaces_register_remote_disabled", method: http.MethodPost, path: "/api/workspaces", body: `{"path":"/tmp"}`, loopback: true},
		restCase{name: "workspaces_register_bad_body", method: http.MethodPost, path: "/api/workspaces", body: `{not json`, loopback: true},
		restCase{name: "workspaces_files_ok", method: http.MethodGet, path: fmt.Sprintf("/api/workspaces/%s/files", wsID), loopback: true},
		restCase{name: "workspaces_files_not_found", method: http.MethodGet, path: "/api/workspaces/nonexistent/files", loopback: true},
		restCase{name: "workspaces_read_ok", method: http.MethodGet, path: fmt.Sprintf("/api/workspaces/%s/file?path=README.md", wsID), loopback: true},
		restCase{name: "workspaces_read_missing_path", method: http.MethodGet, path: fmt.Sprintf("/api/workspaces/%s/file", wsID), loopback: true},
		restCase{name: "workspaces_read_not_found", method: http.MethodGet, path: fmt.Sprintf("/api/workspaces/%s/file?path=nope.txt", wsID), loopback: true},
		restCase{name: "workspaces_raw_ok", method: http.MethodGet, path: fmt.Sprintf("/api/workspaces/%s/raw?path=README.md", wsID), loopback: true},
		restCase{name: "workspaces_search_ok", method: http.MethodGet, path: fmt.Sprintf("/api/workspaces/%s/search?q=hello", wsID), loopback: true},
		restCase{name: "workspaces_write_bad_body", method: http.MethodPost, path: fmt.Sprintf("/api/workspaces/%s/file", wsID), body: `{not json`, loopback: true},
	)

	// --- Events ---
	cases = append(cases,
		restCase{name: "events_list_ok", method: http.MethodGet, path: "/api/events", loopback: true},
		restCase{name: "events_list_unauth", method: http.MethodGet, path: "/api/events", loopback: false},
		restCase{name: "events_session_ok", method: http.MethodGet, path: "/api/events/nonexistent", loopback: true},
	)

	// --- Agents ---
	cases = append(cases,
		restCase{name: "agents_list_ok", method: http.MethodGet, path: "/api/agents", loopback: true},
		restCase{name: "agents_upsert_bad_body", method: http.MethodPost, path: "/api/agents", body: `{not json`, loopback: true},
		restCase{name: "agents_delete_ok", method: http.MethodDelete, path: "/api/agents/fixture-agent", loopback: true},
		restCase{name: "agents_autodetect_ok", method: http.MethodPost, path: "/api/agents/autodetect", loopback: true},
	)

	// --- Sessions ---
	cases = append(cases,
		restCase{name: "sessions_list_ok", method: http.MethodGet, path: "/api/sessions", loopback: true},
		restCase{name: "sessions_get_not_found", method: http.MethodGet, path: "/api/sessions/nonexistent", loopback: true},
		restCase{name: "sessions_export_not_found", method: http.MethodGet, path: "/api/sessions/nonexistent/export", loopback: true},
		restCase{name: "sessions_create_bad_body", method: http.MethodPost, path: "/api/sessions", body: `{not json`, loopback: true},
		restCase{name: "sessions_create_unknown_agent", method: http.MethodPost, path: "/api/sessions", body: `{"agentId":"no-such-agent","modelId":"m","workspaceId":""}`, loopback: true},
		restCase{name: "sessions_patch_bad_body", method: http.MethodPatch, path: "/api/sessions/nonexistent", body: `{not json`, loopback: true},
		restCase{name: "sessions_prompt_not_found", method: http.MethodPost, path: "/api/sessions/nonexistent/prompt", body: `{"content":"hi"}`, loopback: true},
		restCase{name: "sessions_cancel_not_found", method: http.MethodPost, path: "/api/sessions/nonexistent/cancel", loopback: true},
		restCase{name: "sessions_close_not_found", method: http.MethodDelete, path: "/api/sessions/nonexistent", loopback: true},
		restCase{name: "sessions_context_bad_body", method: http.MethodPost, path: "/api/sessions/nonexistent/context", body: `{not json`, loopback: true},
		restCase{name: "sessions_providers_not_found", method: http.MethodGet, path: "/api/sessions/nonexistent/providers", loopback: true},
	)

	// --- Permissions ---
	cases = append(cases,
		restCase{name: "permissions_pending_ok", method: http.MethodGet, path: "/api/permissions/pending", loopback: true},
		restCase{name: "permissions_respond_bad_body", method: http.MethodPost, path: "/api/permissions/nonexistent/respond", body: `{not json`, loopback: true},
	)

	// --- MCP ---
	cases = append(cases,
		restCase{name: "mcp_get_ok", method: http.MethodGet, path: "/api/mcp", loopback: true},
		restCase{name: "mcp_put_bad_body", method: http.MethodPut, path: "/api/mcp", body: `{not json`, loopback: true},
		restCase{name: "mcp_patch_server_bad_body", method: http.MethodPatch, path: "/api/mcp/servers/fixture", body: `{not json`, loopback: true},
		restCase{name: "mcp_status_ok", method: http.MethodGet, path: "/api/mcp/status", loopback: true},
	)

	return cases
}

// runRESTCase executes one restCase against the harness server and returns the
// redacted fixture. It also registers any newly-revealed secrets (e.g. the
// pairing token from /api/pair/initiate) with the harness redactor so later
// captures scrub them.
func runRESTCase(h *harness, c restCase) (restFixture, error) {
	var bodyReader io.Reader
	if c.body != "" {
		bodyReader = strings.NewReader(c.body)
	}
	req := httptest.NewRequest(c.method, c.path, bodyReader)
	if c.loopback {
		req.RemoteAddr = "127.0.0.1:1234"
	} else {
		req.RemoteAddr = "10.0.0.7:1234"
	}
	if c.authHeader != "" {
		req.Header.Set("Authorization", c.authHeader)
	}
	if c.origin != "" {
		req.Header.Set("Origin", c.origin)
	}
	if c.body != "" {
		req.Header.Set("Content-Type", "application/json")
	}

	rec := httptest.NewRecorder()
	h.server.Handler().ServeHTTP(rec, req)

	body := rec.Body.String()
	// Register secrets revealed by pairing responses so subsequent captures
	// (and the WS/CLI captures) scrub them. The pairing session JSON contains
	// "token" and "passcode" fields.
	if strings.Contains(c.path, "/api/pair/") {
		registerPairingSecrets(h, body)
	}
	// Defense-in-depth: scrub any unregistered token/secret/secretHash fields.
	body = ScrubUnregisteredTokens(body)
	body = h.redactor.String(body)

	return restFixture{
		Method:      c.method,
		Path:        c.path,
		Status:      rec.Code,
		ContentType: rec.Header().Get("Content-Type"),
		Body:        body,
	}, nil
}

// registerPairingSecrets parses a pairing response body and registers the
// token and passcode values with the redactor. Failures are non-fatal: a
// malformed body simply means no secrets are registered.
func registerPairingSecrets(h *harness, body string) {
	var s struct {
		Token    string `json:"token"`
		Passcode string `json:"passcode"`
	}
	if err := json.Unmarshal([]byte(body), &s); err != nil {
		return
	}
	if s.Token != "" {
		h.redactor.RegisterSecret(s.Token, "<REDACTED_TOKEN>")
	}
	if s.Passcode != "" {
		h.redactor.RegisterSecret(s.Passcode, "<REDACTED_PASSCODE>")
	}
}

File: tests/contract_runner/rest.rs
//! REST differential tests.
//!
//! Each test case mirrors a case from the Go fixture harness
//! (`tests/contract/go-fixtures/rest.go` → `buildRESTCases`). The runner:
//!
//! 1. Reads the golden fixture (`golden/rest/<name>.json`) to get the expected
//!    response envelope (method, path, status, contentType, body).
//! 2. Substitutes `<REDACTED_WORKSPACE_ID>` in the path with the real workspace
//!    ID discovered at runtime.
//! 3. Makes the HTTP request to the running backend.
//! 4. Redacts the response (same redaction rules as the Go harness).
//! 5. Compares the redacted response against the golden fixture:
//!    - Envelope fields (method, path, status, contentType): exact.
//!    - Body: semantic JSON for objects/arrays, exact bytes for text/errors.
//!
//! Cases that require non-loopback connections (the `_unauth` cases) are
//! skipped — the black-box runner connects via localhost, so the server's
//! loopback auth bypass always applies. These cases are documented as a known
//! limitation in the README.

use std::path::PathBuf;
use std::time::Duration;

use anyhow::{Context, Result};
use reqwest::Method;

use crate::compare;
use crate::harness::BackendHarness;
use crate::redactor::Redactor;

/// A single REST test case. Mirrors `restCase` in go-fixtures/rest.go.
struct RestCase {
    /// The golden fixture name (without extension).
    name: &'static str,
    /// The HTTP method.
    method: &'static str,
    /// The request path (may include query string and <REDACTED_WORKSPACE_ID>).
    path: &'static str,
    /// The request body for POST/PUT/PATCH, or empty for GET/DELETE.
    body: &'static str,
}

/// The full list of REST cases. This mirrors `buildRESTCases` in
/// go-fixtures/rest.go, excluding the `_unauth` cases (which require
/// non-loopback connections and can't be tested black-box).
const REST_CASES: &[RestCase] = &[
    // Health
    RestCase { name: "health_ok", method: "GET", path: "/health", body: "" },
    // Pairing
    RestCase { name: "pair_initiate_ok", method: "POST", path: "/api/pair/initiate", body: r#"{"host":"localhost","port":7337}"# },
    RestCase { name: "pair_initiate_bad_body", method: "POST", path: "/api/pair/initiate", body: r#"{not json"# },
    RestCase { name: "pair_verify_passcode_bad_body", method: "POST", path: "/api/pair/verify-passcode", body: r#"{not json"# },
    RestCase { name: "pair_verify_passcode_wrong", method: "POST", path: "/api/pair/verify-passcode", body: r#"{"passcode":"wrong wrong wrong wrong","deviceName":"dev"}"# },
    RestCase { name: "pair_verify_token_wrong", method: "POST", path: "/api/pair/verify-token", body: r#"{"token":"deadbeef","deviceName":"dev"}"# },
    // Devices
    RestCase { name: "devices_list_ok", method: "GET", path: "/api/devices", body: "" },
    RestCase { name: "devices_revoke_not_found", method: "DELETE", path: "/api/devices/nonexistent", body: "" },
    RestCase { name: "devices_cancel_revocation_bad_body", method: "POST", path: "/api/devices/cancel-revocation", body: r#"{not json"# },
    // Pending actions
    RestCase { name: "pending_actions_list_ok", method: "GET", path: "/api/pending-actions", body: "" },
    RestCase { name: "workspaces_cancel_registration_bad_body", method: "POST", path: "/api/workspaces/cancel-registration", body: r#"{not json"# },
    // Workspaces
    RestCase { name: "workspaces_list_ok", method: "GET", path: "/api/workspaces", body: "" },
    RestCase { name: "workspaces_register_remote_disabled", method: "POST", path: "/api/workspaces", body: r#"{"path":"/tmp"}"# },
    RestCase { name: "workspaces_register_bad_body", method: "POST", path: "/api/workspaces", body: r#"{not json"# },
    RestCase { name: "workspaces_files_ok", method: "GET", path: "/api/workspaces/<REDACTED_WORKSPACE_ID>/files", body: "" },
    RestCase { name: "workspaces_files_not_found", method: "GET", path: "/api/workspaces/nonexistent/files", body: "" },
    RestCase { name: "workspaces_read_ok", method: "GET", path: "/api/workspaces/<REDACTED_WORKSPACE_ID>/file?path=README.md", body: "" },
    RestCase { name: "workspaces_read_missing_path", method: "GET", path: "/api/workspaces/<REDACTED_WORKSPACE_ID>/file", body: "" },
    RestCase { name: "workspaces_read_not_found", method: "GET", path: "/api/workspaces/<REDACTED_WORKSPACE_ID>/file?path=nope.txt", body: "" },
    RestCase { name: "workspaces_raw_ok", method: "GET", path: "/api/workspaces/<REDACTED_WORKSPACE_ID>/raw?path=README.md", body: "" },
    RestCase { name: "workspaces_search_ok", method: "GET", path: "/api/workspaces/<REDACTED_WORKSPACE_ID>/search?pattern=hello", body: "" },
    RestCase { name: "workspaces_write_bad_body", method: "POST", path: "/api/workspaces/<REDACTED_WORKSPACE_ID>/file", body: r#"{not json"# },
    // Events
    RestCase { name: "events_list_ok", method: "GET", path: "/api/events", body: "" },
    RestCase { name: "events_session_ok", method: "GET", path: "/api/events/nonexistent", body: "" },
    // Agents
    RestCase { name: "agents_list_ok", method: "GET", path: "/api/agents", body: "" },
    RestCase { name: "agents_upsert_bad_body", method: "POST", path: "/api/agents", body: r#"{not json"# },
    RestCase { name: "agents_delete_ok", method: "DELETE", path: "/api/agents/fixture-agent", body: "" },
    RestCase { name: "agents_autodetect_ok", method: "POST", path: "/api/agents/autodetect", body: "" },
    // Sessions
    RestCase { name: "sessions_list_ok", method: "GET", path: "/api/sessions", body: "" },
    RestCase { name: "sessions_get_not_found", method: "GET", path: "/api/sessions/nonexistent", body: "" },
    RestCase { name: "sessions_export_not_found", method: "GET", path: "/api/sessions/nonexistent/export", body: "" },
    RestCase { name: "sessions_create_bad_body", method: "POST", path: "/api/sessions", body: r#"{not json"# },
    RestCase { name: "sessions_create_unknown_agent", method: "POST", path: "/api/sessions", body: r#"{"agentId":"no-such-agent","modelId":"m","workspaceId":""}"# },
    RestCase { name: "sessions_patch_bad_body", method: "PATCH", path: "/api/sessions/nonexistent", body: r#"{not json"# },
    RestCase { name: "sessions_prompt_not_found", method: "POST", path: "/api/sessions/nonexistent/prompt", body: r#"{"content":"hi"}"# },
    RestCase { name: "sessions_cancel_not_found", method: "POST", path: "/api/sessions/nonexistent/cancel", body: "" },
    RestCase { name: "sessions_close_not_found", method: "DELETE", path: "/api/sessions/nonexistent", body: "" },
    RestCase { name: "sessions_context_bad_body", method: "POST", path: "/api/sessions/nonexistent/context", body: r#"{not json"# },
    RestCase { name: "sessions_providers_not_found", method: "GET", path: "/api/sessions/nonexistent/providers", body: "" },
    // Permissions
    RestCase { name: "permissions_pending_ok", method: "GET", path: "/api/permissions/pending", body: "" },
    RestCase { name: "permissions_respond_bad_body", method: "POST", path: "/api/permissions/nonexistent/respond", body: r#"{not json"# },
    // MCP
    RestCase { name: "mcp_get_ok", method: "GET", path: "/api/mcp", body: "" },
    RestCase { name: "mcp_put_bad_body", method: "PUT", path: "/api/mcp", body: r#"{not json"# },
    RestCase { name: "mcp_patch_server_bad_body", method: "PATCH", path: "/api/mcp/servers/fixture", body: r#"{not json"# },
    RestCase { name: "mcp_status_ok", method: "GET", path: "/api/mcp/status", body: "" },
];

/// Find a REST case by name from the static list.
fn find_case(name: &str) -> Option<&'static RestCase> {
    REST_CASES.iter().find(|c| c.name == name)
}

/// Resolve the golden fixture directory path.
fn golden_dir(harness: &BackendHarness) -> PathBuf {
    harness.repo_root.join("tests").join("contract").join("golden").join("rest")
}

/// Run a single REST test case by name. Reads the golden fixture, makes the
/// HTTP request, redacts the response, and compares.
pub async fn run_case(harness: &BackendHarness, name: &str) {
    let case = find_case(name).unwrap_or_else(|| panic!("unknown REST case: {name}"));

    // Load the golden fixture.
    let fixture_path = golden_dir(harness).join(format!("{name}.json"));
    let fixture_text = std::fs::read_to_string(&fixture_path)
        .with_context(|| format!("read golden fixture {}", fixture_path.display()))
        .expect("read golden fixture");
    let fixture: serde_json::Value =
        serde_json::from_str(&fixture_text).expect("parse golden fixture JSON");

    // Discover the real workspace ID for path substitution.
    let ws_id = harness.workspace_id().await;

    // Build the redactor with the run's known secrets and paths.
    let mut redactor = Redactor::new();
    redactor.register_path(&harness.state_dir);
    if let Ok(home) = std::env::var("HOME") {
        redactor.register_path(&home);
    }
    redactor.register_path(&harness.repo_root);
    redactor.register_secret(&ws_id, crate::redactor::REDACTED_WORKSPACE_ID);
    // Register the ephemeral port so it is redacted in responses.
    redactor.register_secret(
        &format!("127.0.0.1:{}", harness.port),
        "127.0.0.1:<REDACTED_PORT>",
    );
    redactor.register_secret(
        &format!("localhost:{}", harness.port),
        "localhost:<REDACTED_PORT>",
    );

    // Substitute <REDACTED_WORKSPACE_ID> in the path with the real workspace ID.
    let actual_path = case.path.replace("<REDACTED_WORKSPACE_ID>", &ws_id);

    // Make the HTTP request.
    let method = Method::from_bytes(case.method.as_bytes()).expect("valid HTTP method");
    let url = format!("{}{}", harness.base_url, actual_path);

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(10))
        .build()
        .expect("build reqwest client");

    let mut req = client.request(method, &url);
    if !case.body.is_empty() {
        req = req.body(case.body.to_string())
            .header("Content-Type", "application/json");
    }

    let resp = req.send().await.with_context(|| format!("send request {case.method} {url}"))
        .expect("send HTTP request");

    let status = resp.status().as_u16();
    let content_type = resp
        .headers()
        .get("content-type")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    let body = resp.text().await.expect("read response body");

    // Redact the response body.
    let redacted_body = redactor.redact(&body);

    // Redact the actual path (it contains the real workspace ID which should
    // be replaced with <REDACTED_WORKSPACE_ID> for comparison).
    let redacted_path = redactor.redact(&actual_path);

    // Compare the envelope (method, path, status, contentType).
    if let Err(e) = compare::compare_envelope(
        case.method,
        &redacted_path,
        status,
        &content_type,
        &fixture,
    ) {
        eprintln!("[contract] FAIL: {name}\n{e}");
        panic!("REST case {name} failed: {e}");
    }

    // Compare the body.
    let expected_body = compare::extract_body(&fixture);
    if let Err(e) = compare::compare_body(&redacted_body, expected_body, &content_type) {
        eprintln!("[contract] FAIL: {name}\n{e}");
        panic!("REST case {name} body mismatch: {e}");
    }

    eprintln!("[contract] PASS: {name} ({case.method} {actual_path} → {status})");
}
